### PR TITLE
Improvements for handling merged corporation

### LINKED ISSUE #1913: [18Mex][1846][18LA][1817][Others?] Spreadsheet and Entities tab don't show closed / merged companies
The spreadsheet tab and entities tab doesn't show closed/merged companies. If a company operated for a few ORs and then closed in 1846 it would not show that company in the SS.

### LINKED ISSUE #2003: 18MEX: Merge fails in case the current entity tries to merge into NdM
Consider game 13258, action 308. 
If CHI accepts the game crashes.

I presume this is cause that CHI is deleted from the list of entities, which result in nil.
The handling of merged corporation needs to be improved. Consider using closed, and avoid doing steps for closed corporations, but they can still be shown in entity list, possibly with (CLOSED) after.

### PR DESCRIPTION
**Pull request by perwestling**
Merged corporation is now kept as entity for operating round,
with (CLOSED) after it. If current entity is merged, any remaining
actions is skipped silently.

The merged (closed) corporation are now skipped on entities
and game (stock) page, but remains on the spread sheet tab.

Correct removal of shares from merged corporation in 18MEX.

This fixes the following titles:
- Fixes #2003, the crash was caused by removing of a corporation
  from round entities - caused index error. Now the corporation
  remains but it will not act in the current operation round.
- Fixes #1913 for 18MEX - merged corporation does now remain on
  spread sheet, with all 10 shares in IPO, no share or par price,
  but do show operating history.
  For 1846, 18LA, 1817, Others? it depends on how close of
  corporation is handled.

Added close! and closed? methods to minor and corporation. These
only set closed attribute, sets owner to bank, and set floated to
false. Any clean up needed have to be done in the title that calls
the close! method.

### CODE REVIEW

File: assets/app/view/game/entities.rb [hunk 1]
```diff
@@ -17,13 +17,13 @@ def render
...
         }
-
         players = @game.players
         if (i = players.map(&:name).rindex(@user&.dig(:name)))
           players = players.rotate(i)
         end
 
-        player_owned, bank_owned = (@game.corporations + @game.minors).sort_by(&:name).partition(&:owner)
+        entity_candidates = (@game.corporations + @game.minors).reject(&:closed?)
+        player_owned, bank_owned = entity_candidates.sort_by(&:name).partition(&:owner)
```
Comment (tobymao) [inline assets/app/view/game/entities.rb:26]: you can just do this all in one line

x, y = (...)
.rject(&:closed?)
.sort_by
.partition

File: assets/app/view/game/spreadsheet.rb [hunk 2]
```diff
@@ -270,6 +270,14 @@ def render_corporation(corporation, operating_order, current_round, index)
         ])
       end
 
+      def par_price?(corporation)
```
Comment (tobymao) [inline assets/app/view/game/spreadsheet.rb:273]: should you just put the closed in the corporation getter intsead?

File: assets/app/view/game/stock_market.rb [hunk 3]
```diff
@@ -145,7 +145,7 @@ def grid_2d
         @game.stock_market.market.flat_map do |row_prices|
           row = row_prices.map do |price|
             if price
-              corporations = price.corporations
+              corporations = price.corporations.reject(&:closed?)
```
Comment (tobymao) [inline assets/app/view/game/stock_market.rb:148]: if a corporation is closed they shouldn't be in the price anyways, so i don't know if we should do this

File: lib/engine/corporation.rb [hunk 4]
```diff
@@ -213,5 +214,15 @@ def transfer(ownable_type, to)
       ownables.clear
       transferred
     end
+
+    def closed?
+      @closed
+    end
+
+    def close!(bank)
```
Comment (tobymao) [inline lib/engine/corporation.rb:222]: maybe here we sohuld remove the corporation from the share price

File: lib/engine/game/g_18_mex.rb [hunk 5]
```diff
@@ -502,6 +486,11 @@ def replace_token(major, major_token, exchange_tokens)
         city.place_token(ndm, ndm_replacement, check_tokenable: false)
         exchange_tokens.delete(ndm_replacement)
       end
+
+      def close_major(major)
+        major.ipoed = false
```
Comment (tobymao) [inline lib/engine/game/g_18_mex.rb:491]: shoudln't this be done in close! ?

File: lib/engine/minor.rb [hunk 6]
```diff
@@ -3,10 +3,10 @@
 require_relative 'assignable'
 require_relative 'abilities'
 require_relative 'entity'
-require_relative 'ownable'
 require_relative 'passer'
 require_relative 'spender'
 require_relative 'operator'
+require_relative 'ownable'
```
Comment (tobymao) [inline lib/engine/minor.rb:9]: why is this moved?

File: lib/engine/corporation.rb (hunk 4 above)
Comment (tobymao) [reply lib/engine/corporation.rb:222]: why pass bank here? just set owner to be nil

File: lib/engine/minor.rb [hunk 7]
```diff
@@ -57,5 +58,15 @@ def float!
     def inspect
       "<#{self.class.name}: #{id}>"
     end
+
+    def closed?
+      @closed
+    end
+
+    def close!(bank)
```
Comment (tobymao) [inline lib/engine/minor.rb:66]: just set owner nil

File: assets/app/view/game/entities.rb (hunk 1 above)
Comment (perwestling) [reply assets/app/view/game/entities.rb:26]: Fixed

File: assets/app/view/game/spreadsheet.rb (hunk 2 above)
Comment (perwestling) [reply assets/app/view/game/spreadsheet.rb:273]: Ok, have now put some logic in the par_price and share_price getters.
Had to change accessor method for share_price

File: assets/app/view/game/stock_market.rb (hunk 3 above)
Comment (perwestling) [reply assets/app/view/game/stock_market.rb:148]: I added a close(corporation) method in stock_market that removes it from price. That works.

File: lib/engine/corporation.rb (hunk 4 above)
Comment (perwestling) [reply lib/engine/corporation.rb:222]: Have added a deletion of share price from stock market. Now I pass game instead. And set owner to nil.

File: lib/engine/game/g_18_mex.rb (hunk 5 above)
Comment (perwestling) [reply lib/engine/game/g_18_mex.rb:491]: Ok, moved.

File: lib/engine/minor.rb (hunk 6 above)
Comment (perwestling) [reply lib/engine/minor.rb:9]: I think I moved it to have it after operator. Have now moved them so they are alphabetical.

**Comment by perwestling** [comment]
@tobymao Now I have pushed an updated version of this, with your review comments handled.

The only thing that I wonder about is if the cleanup code should be moved to general code?
It could be:

1. Remove tokens on map
2. Delete all trains in entity
3. Move money in entity to bank
4. Move shares from players to IPO

Then it is up to the game title to transfer stuff, give rebate, swap tokens etc if it needs to do things.


-

File: lib/engine/stock_market.rb [hunk 8]
```diff
@@ -117,6 +117,10 @@ def move(corporation, row, column, force: false)
       share_price.corporations << corporation
     end
 
+    def close(corporation)
```
Comment (tobymao) [inline lib/engine/stock_market.rb:120]: delete this

File: lib/engine/corporation.rb [hunk 9]
```diff
@@ -213,5 +222,17 @@ def transfer(ownable_type, to)
       ownables.clear
       transferred
     end
+
+    def closed?
+      @closed
+    end
+
+    def close!(game)
+      game.stock_market.close(self)
```
Comment (tobymao) [inline lib/engine/corporation.rb:231]: you don't need to pass the game,

just do share_price.corporations.delete(self)

File: lib/engine/minor.rb (hunk 7 above)
Comment (perwestling) [reply lib/engine/minor.rb]: I corrected that, but forgot to comment. So it has been fixed.

File: lib/engine/stock_market.rb (hunk 8 above)
Comment (perwestling) [reply lib/engine/stock_market.rb]: ok

File: lib/engine/corporation.rb (hunk 9 above)
Comment (perwestling) [reply lib/engine/corporation.rb]: ok

Added & as the share price might not have been set (in case of merging a corporation that has never been bought shares from),

**Comment by perwestling** [comment]
@tobymao Another PR that has been updated.

**Review by tobymao — APPROVED** [review]
